src/containers/MediaElementContainer.js

```diff
@@ -2,81 +2,81 @@ import React, { Component } from 'react';
 import MediaElement from '../components/MediaElement';
 import PropTypes from 'prop-types';
 import ErrorMessage from '../components/ErrorMessage';
-import manifesto from 'manifesto.js';
+import {
+  getChoiceItems,
+  getSources,
+  getMediaType
+} from '../services/iiif-parser';
+import { connect } from 'react-redux';
 
 class MediaElementContainer extends Component {
   state = {
     manifest: this.props.manifest,
     ready: false,
     sources: [],
     mediaType: null,
+    canvasIndex: 0,
     error: null
   };
 
   componentDidMount() {
     const { manifest } = this.state;
-    let choiceItems = [];
-
-    try {
-      choiceItems = manifesto
-        .create(manifest)
-        .getSequences()[0]
-        .getCanvases()[0]
-        .getContent()[0]
-        .getBody();
-    } catch (e) {}
-
-    this.prepSources(choiceItems);
-  }
-
-  getSources(choiceItems) {
-    const sources = choiceItems.map(item => {
-      return {
-        src: item.id,
-        // TODO: make type more generic, possibly use mime-db
-        format: item.getFormat().value
-      };
-    });
-    return sources;
-  }
-
-  getMediaType(choiceItems) {
-    let allTypes = choiceItems.map(item => item.getType().value);
-    let uniqueTypes = allTypes.filter((t, index) => {
-      return allTypes.indexOf(t) === index;
-    });
-    if (uniqueTypes.length === 1) {
-      return uniqueTypes[0];
-    }
-    // Default type if there are different types
-    return 'audio';
-  }
-
-  prepSources(choiceItems) {
+    // Get the first canvas in manifest
+    let choiceItems = getChoiceItems(manifest, 0);
     if (choiceItems.length === 0) {
       this.setState({
         error: 'No media choice items found in manifest'
       });
       return;
     }
 
-    const sources = this.getSources(choiceItems);
-    const mediaType = this.getMediaType(choiceItems);
+    const sources = getSources(choiceItems);
+    const mediaType = getMediaType(choiceItems);
     this.setState({
       ready: true,
       sources,
       mediaType
     });
   }
 
+  static getDerivedStateFromProps(nextProps, prevState) {
+    const { reload, canvasIndex } = nextProps;
+    const { manifest } = prevState;
+    if (reload) {
+      const choiceItems = getChoiceItems(manifest, canvasIndex);
+      if (choiceItems.length === 0) {
+        return {
+          error: 'No media choice items found in manifest'
+        };
+      }
+      const sources = getSources(choiceItems);
+      const mediaType = getMediaType(choiceItems);
+      return {
+        sources,
+        mediaType,
+        canvasIndex,
+        ready: true
+      };
```

Comment:
It looks like this matches what is in `componentDidMount`.  Does it make sense to dry this up by extracting a method that is shared by both?

Reply:
The static method cannot call functions within the class. Maybe I can introduce a method in `iiif-parser` with all this and call it in both `componentDidMount` and `getDerivedStateFromProps` functions?

Reply:
If it makes it more complex then maybe it doesn't make sense to dry it up.

Reply:
No it's not. I was able to get `sources`, `mediaType` from one single method in iiif-parser, which we have duplicated here. And I was able to remove `componentDidMount` function and make the `getDerivedStateFromProps` function work for all scenarios. I think it looks better now.